### PR TITLE
Added unit tests with mock client , RBAC error handling, README docs and DSCI namespace auto-discovery in mcp server

### PR DESCRIPTION
**Pull request by harsh1947-jain**
## Description
Implements unit tests for all 7 tools using mock clients, and integration test documentation. Handles RBAC failures, missing CRDs, and empty namespaces gracefully. Supports namespace auto-discovery from DSCI with flag overrides. Includes README with setup guide, tool reference, example outputs, and MCP config examples for Cursor and Claude Desktop.

https://redhat.atlassian.net/browse/RHOAIENG-56937

## How Has This Been Tested?

Unit Tests

Unit tests for all 7 tools using mock Kubernetes clients, covering RBAC failures, missing CRDs and empty namespaces.

## Screenshot or short clip
image

## Merge criteria

- [X] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [X] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [X] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [X] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [X] The developer has manually tested the changes and verified that the changes work

### E2E test suite update requirement

When bringing new changes to the operator code, such changes are by default required to be accompanied by extending and/or updating the E2E test suite accordingly.

To opt-out of this requirement:
1. **Please inspect the [opt-out guidelines](https://github.com/opendatahub-io/opendatahub-operator/blob/main/docs/e2e-update-requirement-guidelines.md)**, to determine if the nature of the PR changes allows for skipping this requirement
2. If opt-out is applicable, provide justification in the dedicated `E2E update requirement opt-out justification` section below
3. Check the checkbox below:
- [X] Skip requirement to update E2E test suite for this PR
4. Submit/save these changes to the PR description. This will automatically trigger the check.

#### E2E update requirement opt-out justification
Add E2E tests if required.

## Summary by CodeRabbit

* **Documentation**
  * Added comprehensive README and integration test guide with usage examples, setup, and expected outputs

* **New Features**
  * Auto-discovery of application and operator namespaces when omitted
  * Make target to surface integration test instructions

* **Bug Fixes / Improvements**
  * Clearer RBAC and “CRD not installed” error messages
  * More precise pod-log error reporting for invalid requests and permission denials
  * Tools return structured, non-crashing error results

* **Tests**
  * Expanded tests for RBAC failures, missing CRDs, error propagation, and namespace discovery

### CODE REVIEW

File: cmd/mcp-server/helpers_test.go [hunk 1]
```diff
@@ -73,3 +79,76 @@ func TestStringParam(t *testing.T) {
...
+}
+
+func fakeClient(dsci *unstructured.Unstructured) client.Client {
+	s := runtime.NewScheme()
+	_ = scheme.AddToScheme(s)
+	b := fake.NewClientBuilder().WithScheme(s)
+	if dsci != nil {
+		b = b.WithRuntimeObjects(dsci)
+	}
+	return b.Build()
```
Comment (den-rgb) [inline cmd/mcp-server/helpers_test.go:132]: add RBAC-forbidden and CRD-not-installed cases to TestDiscoverAppsNamespace using newForbiddenClient() / newNoMatchClient() as the client, asserting wantErr: true, currently only happy paths are covered for a function that now returns typed errors

File: cmd/mcp-server/tool_platform_health_test.go [hunk 2]
```diff
@@ -96,3 +98,121 @@ func TestPlatformHealth_NilClient(t *testing.T) {
...
+					if !want[s] {
+						t.Errorf("unexpected section %q", s)
+					}
+				}
+			}
+		})
+	}
+}
+
+func TestPlatformHealth_ErrorClients(t *testing.T) {
```
Comment (den-rgb) [inline cmd/mcp-server/tool_platform_health_test.go:201]: TestPlatformHealth_ErrorClients passes explicit namespace args, bypassing discoverAppsNamespace. add a case with empty arguments: map[string]any{} and an error client to cover the "namespace discovery failed" handler response path

File: cmd/mcp-server/helpers.go [hunk 3]
```diff
@@ -65,6 +74,44 @@ func getEnvDefault(key, fallback string) string {
...
+		}
+	}
+
+	if len(list.Items) > 0 {
+		if ns, found, _ := unstructured.NestedString(list.Items[0].Object, "spec", "applicationsNamespace"); found && ns != "" {
+			return ns, nil
+		}
+	}
+
+	return fallback, nil
```
Comment (den-rgb) [inline cmd/mcp-server/helpers.go:133]: discoverOperatorNamespace takes (ctx, client) but ignores both, either simplify signature to discoverOperatorNamespace() string or document why params are reserved

File: cmd/mcp-server/tool_recent_events.go
Comment (den-rgb) [inline cmd/mcp-server/tool_recent_events.go:92]: discoverODHNamespaces re-implements the DSCI list+read from discoverAppsNamespace in helpers.go. extract a shared DSCI lookup helper to avoid divergence

File: cmd/mcp-server/README.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+| `tail_lines` | number | no | `100` | Number of lines from the end of the log |
+
+```jsonc
+// Example call
+{"jsonrpc":"2.0","id":1,"method":"tools/call","params":{"name":"pod_logs","arguments":{
+  "pod_name":"odh-dashboard-abc12","namespace":"opendatahub","tail_lines":10
+}}}
+```
+
+```
```
[the remark this answers is not in the document]
Comment (den-rgb) [reply cmd/mcp-server/README.md]: +1

File: cmd/mcp-server/error_helpers_test.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+				return err
+			},
+			List: func(_ context.Context, _ client.WithWatch, _ client.ObjectList, _ ...client.ListOption) error {
+				return err
+			},
+		}).
+		Build()
+}
+
+func newForbiddenClient() client.Client {
```
Comment (den-rgb) [inline cmd/mcp-server/error_helpers_test.go:30]: np: newForbiddenClient hardcodes group/resource as pods but is used for testing DSCI/Deployment/Event operations, this could be confusing when debugging, you could use `schema.GroupResource{Resource: "resources"}` or parameterize it

File: cmd/mcp-server/helpers_test.go (hunk 1 above)
Comment (harsh1947-jain) [reply cmd/mcp-server/helpers_test.go:132]: Done

File: cmd/mcp-server/helpers.go (hunk 3 above)
Comment (harsh1947-jain) [reply cmd/mcp-server/helpers.go:133]: Simplified the signature.

File: cmd/mcp-server/tool_recent_events.go
Comment (harsh1947-jain) [reply cmd/mcp-server/tool_recent_events.go:92]: Done.

File: cmd/mcp-server/error_helpers_test.go (hunk 5 above)
Comment (harsh1947-jain) [reply cmd/mcp-server/error_helpers_test.go:30]: I have used schema.GroupResource{Resource: "resources"}.

File: cmd/mcp-server/tool_platform_health_test.go (hunk 2 above)
Comment (harsh1947-jain) [reply cmd/mcp-server/tool_platform_health_test.go:201]: Added the test case.

File: cmd/mcp-server/tool_component_status.go [hunk 6]
```diff
@@ -20,17 +22,34 @@ func registerComponentStatus(s *server.MCPServer, kubeClient client.Client) {
...
 			stringParam(req, "component", ""),
-			stringParam(req, "applications_namespace",
-				getEnvDefault(envApplicationsNamespace, defaultAppsNS)),
+			appsNS,
 		)
 		if err != nil {
-			return mcp.NewToolResultError(err.Error()), nil
+			switch {
+			case k8serr.IsForbidden(err):
+				return mcp.NewToolResultError(fmt.Sprintf("RBAC insufficient: %v", err)), nil
```
Comment (den-rgb) [inline cmd/mcp-server/tool_component_status.go]: the RBAC case "RBAC insufficient: %v" forwards raw k8s authz text to MCP consumers. use the same pattern as tool_pod_logs.go

File: cmd/mcp-server/tool_component_status.go [hunk 7]
```diff
@@ -20,17 +22,34 @@ func registerComponentStatus(s *server.MCPServer, kubeClient client.Client) {
...
+			case meta.IsNoMatchError(err):
+				return mcp.NewToolResultError(fmt.Sprintf(
+					"CRD not installed: component %q requires a CRD that is not registered on this cluster",
+					stringParam(req, "component", ""))), nil
+			default:
```
Comment (den-rgb) [inline cmd/mcp-server/tool_component_status.go:97]: return a generic "failed to determine component status" message and log.Printf the real error, same approach you used in discoverAppsNamespace

File: cmd/mcp-server/helpers.go [hunk 8]
```diff
@@ -65,6 +81,47 @@ func getEnvDefault(key, fallback string) string {
...
+	list := &unstructured.UnstructuredList{}
+	list.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   clusterhealth.DSCInitializationGVK.Group,
+		Version: clusterhealth.DSCInitializationGVK.Version,
+		Kind:    clusterhealth.DSCInitializationGVK.Kind + "List",
+	})
+	if err := kubeClient.List(ctx, list); err != nil {
+		switch {
+		case k8serr.IsForbidden(err):
+			log.Printf("DSCI discovery: RBAC forbidden: %v", err)
```
Comment (den-rgb) [inline cmd/mcp-server/helpers.go:111]: wrap the original error: fmt.Errorf("failed to discover DSCI namespace: %w", err)

File: cmd/mcp-server/tool_component_status.go [hunk 9]
```diff
@@ -20,17 +22,34 @@ func registerComponentStatus(s *server.MCPServer, kubeClient client.Client) {
 		mcp.WithString("component", mcp.Required(),
 			mcp.Description("Component name, e.g. kserve, dashboard, workbenches")),
 		mcp.WithString("applications_namespace",
-			mcp.Description("Apps namespace. Default: opendatahub")),
+			mcp.Description("Apps namespace. Auto-discovered from DSCI if not provided. Falls back to env var or 'opendatahub'.")),
```
Comment (den-rgb) [inline cmd/mcp-server/tool_component_status.go]: update the description to match classify_failure and platform_health

File: cmd/mcp-server/tool_recent_events_test.go [hunk 10]
```diff
@@ -112,6 +113,32 @@ func TestRecentEvents_SortOrder(t *testing.T) {
 	}
 }
 
+func TestRecentEvents_ErrorClients(t *testing.T) {
```
Comment (den-rgb) [inline cmd/mcp-server/tool_recent_events_test.go:144]: these error-client tests pass explicit Namespaces, bypassing discoverODHNamespaces. add a handler-level test with empty args + error client (like TestPlatformHealth_ErrorClients's "namespace discovery failed" case) to cover the discovery path

File: cmd/mcp-server/INTEGRATION_TEST.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+## Setup
+
+Run all commands from the repository root. Add these helpers to your shell:
+
+```bash
+# For tools returning JSON
+call_tool() {
+  echo "{\"jsonrpc\":\"2.0\",\"id\":1,\"method\":\"tools/call\",\"params\":{\"name\":\"$1\",\"arguments\":${2:-\{\}}}}" \
+    | ./bin/mcp-server 2>/dev/null | jq -r '.result.content[0].text' | jq .
```
Comment (den-rgb) [inline cmd/mcp-server/INTEGRATION_TEST.md]: remove 2>/dev/null from the helper commands — it suppresses the RBAC/CRD diagnostic messages this PR introduces

File: cmd/mcp-server/tool_component_status.go (hunk 6 above)
Comment (harsh1947-jain) [reply cmd/mcp-server/tool_component_status.go]: Done.

File: cmd/mcp-server/tool_component_status.go (hunk 7 above)
Comment (harsh1947-jain) [reply cmd/mcp-server/tool_component_status.go:97]: Done.

File: cmd/mcp-server/helpers.go (hunk 8 above)
Comment (harsh1947-jain) [reply cmd/mcp-server/helpers.go:111]: Done.

File: cmd/mcp-server/tool_component_status.go [hunk 12]
```diff
@@ -66,17 +70,46 @@ func registerComponentStatus(s *server.MCPServer, kubeClient client.Client) {
...
+		}
 
-		result, err := clusterhealth.GetComponentStatus(ctx, kubeClient, component, appsNS)
+		result, err := clusterhealth.GetComponentStatus(ctx, kubeClient,
+			stringParam(req, "component", ""),
+			appsNS,
+		)
 		if err != nil {
-			return mcp.NewToolResultError(err.Error()), nil
+			component := stringParam(req, "component", "")
```
Comment (den-rgb) [inline cmd/mcp-server/tool_component_status.go]: code doesn't compile: component is declared inside the if err != nil block at L99 but used at L115 outside it. extract component := stringParam(req, "component", "") before the GetComponentStatus call so it's in function scope

Please verify code works

File: cmd/mcp-server/tool_recent_events_test.go [hunk 13]
```diff
@@ -113,6 +116,50 @@ func TestRecentEvents_SortOrder(t *testing.T) {
...
+				} `json:"result"`
+			}
+			if err := json.Unmarshal(raw, &rpc); err != nil {
+				t.Fatalf("unmarshal rpc: %v", err)
+			}
+			if len(rpc.Result.Content) == 0 {
+				t.Fatal("empty content")
+			}
+			if !strings.Contains(rpc.Result.Content[0].Text, tt.wantInErr) {
+				t.Errorf("error text=%q, want substring %q", rpc.Result.Content[0].Text, tt.wantInErr)
```
Comment (den-rgb) [inline cmd/mcp-server/tool_recent_events_test.go:187]: missing closing braces

please verify code works before pushing

**Comment by den-rgb** [comment]
the discovery-error-to-tool-result mapping is copy-pasted in 4 tools. extract a shared discoveryErrorResult(toolName, err) helper.

File: cmd/mcp-server/tool_classify_failure_test.go [hunk 14]
```diff
@@ -3,10 +3,36 @@ package main
...
+		{"RBAC forbidden", newForbiddenClient()},
+		{"CRD not installed", newNoMatchClient()},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			report := callTool(t, tt.client, nil)
+			fc := failureclassifier.Classify(&report)
+
+			if fc.Category == "" {
```
Comment (den-rgb) [inline cmd/mcp-server/tool_classify_failure_test.go]: both RBAC and CRD cases pass with any non-empty category. add per-case expected values for fc.Category / fc.Evidence

File: cmd/mcp-server/tool_platform_health_test.go [hunk 15]
```diff
@@ -97,6 +99,150 @@ func TestPlatformHealth_NilClient(t *testing.T) {
...
+	tests := []struct {
+		name      string
+		client    client.Client
+		args      map[string]any
+		wantInErr string
+	}{
+		{"RBAC forbidden", newForbiddenClient(), map[string]any{
+			"applications_namespace": "opendatahub",
+			"operator_namespace":     "opendatahub-operator-system",
+		}, "forbidden"},
```
Comment (den-rgb) [inline cmd/mcp-server/tool_platform_health_test.go]: np: assert stable tool-level messages ("RBAC insufficient", "CRD not installed") instead of raw apimachinery strings ("forbidden", "no matches for kind")

File: cmd/mcp-server/helpers.go
Comment (den-rgb) [inline cmd/mcp-server/helpers.go]: trim before checking: if v := strings.TrimSpace(os.Getenv(key)); v != "" { return v }

File: cmd/mcp-server/tool_recent_events.go
Comment (den-rgb) [inline cmd/mcp-server/tool_recent_events.go]: recent_events error: %v leaks raw k8s error text when namespace is explicitly provided. apply the same RBAC/CRD/generic three-way switch you use in tool_component_status.go

Comment (den-rgb) [inline cmd/mcp-server/tool_recent_events.go]: same here

File: cmd/mcp-server/helpers.go
Comment (harsh1947-jain) [reply cmd/mcp-server/helpers.go]: Done

File: cmd/mcp-server/tool_recent_events.go
Comment (harsh1947-jain) [reply cmd/mcp-server/tool_recent_events.go]: Applied the three way switch.

File: cmd/mcp-server/tool_platform_health_test.go [hunk 16]
```diff
@@ -126,8 +128,141 @@ func TestSummarizeReport(t *testing.T) {
 			}
 			if sec.Status != tt.wantStatus {
 				t.Errorf("section %q status = %q, want %q", tt.checkSection, sec.Status, tt.wantStatus)
+				}
```
Comment (den-rgb) [inline cmd/mcp-server/tool_platform_health_test.go]: run `make fmt` brackets wrong

File: cmd/mcp-server/tool_recent_events.go [hunk 17]
```diff
@@ -61,19 +62,31 @@ func registerRecentEvents(s *server.MCPServer, kubeClient client.Client) {
...
 			}
 			cfg.Namespaces = discovered
 		}
 
 		events, eventsErr := clusterhealth.RunRecentEvents(ctx, cfg)
 		if eventsErr != nil && len(events) == 0 {
-			return mcp.NewToolResultError(fmt.Sprintf("recent_events error: %v", eventsErr)), nil
+			errMsg := eventsErr.Error()
+			switch {
+			case k8serr.IsForbidden(eventsErr) || strings.Contains(errMsg, "forbidden"):
```
Comment (den-rgb) [inline cmd/mcp-server/tool_recent_events.go]: drop the strings.Contains(errMsg, "forbidden") / "no matches for kind" fallbacks, the typed checks (k8serr.IsForbidden, meta.IsNoMatchError) are sufficient and the string matching can false-positive on unrelated messages.

File: cmd/mcp-server/tool_recent_events.go [hunk 18]
```diff
@@ -61,19 +62,31 @@ func registerRecentEvents(s *server.MCPServer, kubeClient client.Client) {
...
+				return mcp.NewToolResultError(
+					"RBAC insufficient: the operator service-account lacks permission to list events"), nil
+			case meta.IsNoMatchError(eventsErr) || strings.Contains(errMsg, "no matches for kind"):
```
Comment (den-rgb) [inline cmd/mcp-server/tool_recent_events.go]: same as above

File: cmd/mcp-server/tool_recent_events.go
Comment (den-rgb) [inline cmd/mcp-server/tool_recent_events.go]: partial-failure branch still leaks raw error via %v. apply the same sanitization or use "warning: partial failure retrieving events" and log the original error server-side.

File: cmd/mcp-server/tool_recent_events.go (hunk 17 above)
Comment (harsh1947-jain) [reply cmd/mcp-server/tool_recent_events.go]: Dropped the strings.

File: cmd/mcp-server/tool_recent_events.go [hunk 19]
```diff
@@ -61,24 +63,36 @@ func registerRecentEvents(s *server.MCPServer, kubeClient client.Client) {
...
+				return discoveryErrorResult("recent_events", discErr), nil
 			}
 			cfg.Namespaces = discovered
 		}
 
 		events, eventsErr := clusterhealth.RunRecentEvents(ctx, cfg)
 		if eventsErr != nil && len(events) == 0 {
-			return mcp.NewToolResultError(fmt.Sprintf("recent_events error: %v", eventsErr)), nil
+			switch {
+			case k8serr.IsForbidden(eventsErr):
```
Comment (den-rgb) [inline cmd/mcp-server/tool_recent_events.go]: k8serr.IsForbidden and meta.IsNoMatchError are unreachable here, RunRecentEvents joins per-namespace errors into a plain errors.New(strings.Join(...)) in pkg/clusterhealth/events.go:159, discarding the typed error. both RBAC/CRD cases at tool_recent_events_test.go:151-152 expect "failed to retrieve recent events"

File: cmd/mcp-server/tool_pod_logs.go
Comment (den-rgb) [inline cmd/mcp-server/tool_pod_logs.go]: this needs to be sanatised too, go through all instances where still include %v with raw k8s error text.

**Review by den-rgb — APPROVED** [review]
/lgtm
